### PR TITLE
Ostrich test suite for ServiceCache

### PR DESCRIPTION
**Pull request by bivasdas**

### CODE REVIEW

File: perf-test-suite/README.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
+# Performance Test Suite for Ostrich
+
+## Objective
+
+The point of this performance test is to be able to recreate Ostrich performance problems that the 
+ Concierge team ran into in Production. With the eventual aim of being able to fix those performance
+ problems.
+
+Anecdotally, Concierge would run fine as long as the total number of requests-per-second to an Ostrich 
+ service pool was low. As Concierge traffic went up, the Ostrich service pool traffic would went up,
```
Comment (ChrisBono) [inline perf-test-suite/README.md]: "As Conceirge traffic increased, the Ostrich service pool traffic increased,"

File: perf-test-suite/README.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
+# Performance Test Suite for Ostrich
+
+## Objective
+
+The point of this performance test is to be able to recreate Ostrich performance problems that the 
+ Concierge team ran into in Production. With the eventual aim of being able to fix those performance
```
Comment (dump247) [inline perf-test-suite/README.md]: Be aware that this is a public, open source project. Please don't reference internal projects or concerns as not everyone will be familiar with them.

File: perf-test-suite/src/main/java/com/bazaarvoice/ostrich/perftest/core/SimpleServiceFactory.java [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    public String getServiceName() {
+        return "SimpleService";
+    }
+
+    @Override
+    public void configure(ServicePoolBuilder<Service<String, String>> servicePoolBuilder) { }
+
+    @Override
+    public Service<String, String> create(final ServiceEndPoint serviceEndPoint) {
+        final HashFunction hashFunction = HashFunction.valueOf(serviceEndPoint.getId());
```
Comment (ChrisBono) [inline perf-test-suite/src/main/java/com/bazaarvoice/ostrich/perftest/core/SimpleServiceFactory.java]: How does serviceEndPoint.getId() relate to one of the names of the enums SHA5 for instance?

File: perf-test-suite/pom.xml [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    <dependencies>
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>com.bazaarvoice.ostrich</groupId>
+            <artifactId>ostrich-core</artifactId>
+            <version>1.7.1</version>
```
Comment (dump247) [inline perf-test-suite/pom.xml]: use ${project.version} to test the current version of ostrich

File: perf-test-suite/src/main/java/com/bazaarvoice/ostrich/perftest/utils/LoadRunner.java [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                }
+            }
+            shouldContinue = false;
+        } else {
+            int total = workers.size();
+            int done = 0;
+            for (Thread t : workers) {
+                if (!t.isAlive()) done++;
+            }
+            shouldContinue = !(done == total);
```
Comment (ChrisBono) [inline perf-test-suite/src/main/java/com/bazaarvoice/ostrich/perftest/utils/LoadRunner.java]: shouldContinue = (done != total)

File: perf-test-suite/pom.xml [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+            <artifactId>commons-lang3</artifactId>
+            <version>3.2</version>
+        </dependency>
+        <dependency>
+            <groupId>commons-codec</groupId>
+            <artifactId>commons-codec</artifactId>
+            <version>1.8</version>
+        </dependency>
+        <dependency>
+            <groupId>com.yammer.metrics</groupId>
```
Comment (dump247) [inline perf-test-suite/pom.xml]: ostrich is using codahale metrics, so use that instead of yammer metrics

Comment (bivasdas) [inline perf-test-suite/pom.xml]: I tried that. The new metric does not give mean() on timer, just meanRate() and there was some other (deal) breaking changes as well. Hence I used the yammer (older one), at least for now.

File: perf-test-suite/pom.xml (hunk 4 above)
Comment (bivasdas) [inline perf-test-suite/pom.xml]: Its intended to "test" any published version of ostrich, not just the latest release.

File: perf-test-suite/README.md (hunk 2 above)
Comment (bivasdas) [inline perf-test-suite/README.md]: We can rephrase the readme to address with more generic statements. wip.

File: perf-test-suite/pom.xml (hunk 4 above)
Comment (dump247) [inline perf-test-suite/pom.xml]: In that case, you should make it configurable and default to the current version. Testing an old build is unexpected.

File: perf-test-suite/README.md [hunk 7]
```diff
@@ -0,0 +1,194 @@
+# Performance Test Suite for Ostrich
+
+## Objective
+
+The point of this performance test is to be able to recreate Ostrich performance problems observed in 
+ production, with the eventual aim of being able to fix those performance problems.
+
+Anecdotally, services would run fine as long as the total number of requests-per-second to an Ostrich 
+ service pool was low. As traffic increased, the Ostrich service pool traffic would went up,
```
Comment (milosimpson) [inline perf-test-suite/README.md]: As the requests per second to the Ostrich Service Pool increased, the performance of Ostrich (find me a client to use) would degrade, sometimes catastrophically.

File: perf-test-suite/README.md [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+Anecdotally, services would run fine as long as the total number of requests-per-second to an Ostrich 
+ service pool was low. As traffic increased, the Ostrich service pool traffic would went up, 
+ and eventually the performance of the Ostrich based calls would degrade, sometimes catastrophically.
+
+This test has many parameters that can be adjusted
+
+  * The requests-per-second to Ostrich
+  * What work the "Ostrich client" does for each request
+  * Ostrich "pool" parameters (How and when Ostrich will make new Client objects)
```
Comment (milosimpson) [inline perf-test-suite/README.md]: How and when Ostrich will request new Client Objects for each Endpoint

File: perf-test-suite/README.md [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+Note this test is specifically testing a "backend" component of Ostrich (the ServiceCache) that we've 
+ identified as a problem. This test is not testing Ostrich as a "whole" or as a black box.
```
Comment (milosimpson) [inline perf-test-suite/README.md]: This does not test Ostrich as a whole ...

File: perf-test-suite/README.md [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+
+## Notes on Compilation
+
+The suite also takes is a parameter "ostrich.core.version" at compile time to determine which version of 
+ ostrich core should be tested.
+
+    mvn -q clean compile assembly:single -Dostrich.core.version=1.9.0
+
+However, due to breaking changes between versions, i.e. 1.8 & 1.9 uses different version of metrics and
+ subsequently different constructor of ServiceCache, this `may' require you to make changes in the test
```
Comment (milosimpson) [inline perf-test-suite/README.md]: weird single quote marks around the "may"

File: perf-test-suite/README.md [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+    mvn -q clean compile assembly:single -Dostrich.core.version=1.9.0
+
+However, due to breaking changes between versions, i.e. 1.8 & 1.9 uses different version of metrics and
+ subsequently different constructor of ServiceCache, this `may' require you to make changes in the test 
+ for successful compilation.
+
+
+## Running The Suite
+
+The test suite takes various parameters, such as to determine the load on the cache (# of threads), the load
```
Comment (milosimpson) [inline perf-test-suite/README.md]: To determine the performance overhead of Ostrich under varying loads, this test suite takes various paramters such as ...

File: perf-test-suite/src/main/java/com/bazaarvoice/ostrich/pool/ServiceRunner.java [hunk 12]
```diff
@@ -0,0 +1,222 @@
...
+
+            String work = serviceEndPoint.getPayload();
+            String result = service.process(work);
+
+            TimerContext checkinTimeContext = _checkinTimer.time();
+            _serviceCache.checkIn(serviceHandle);
+            checkinTimeContext.stop();
+
+            _serviceMeter.mark();
+            totalTimeContext.stop();
```
Comment (milosimpson) [inline perf-test-suite/src/main/java/com/bazaarvoice/ostrich/pool/ServiceRunner.java]: totalTimeContext.stop() should be in a finally

File: perf-test-suite/src/main/java/com/bazaarvoice/ostrich/pool/ServiceRunner.java [hunk 13]
```diff
@@ -0,0 +1,222 @@
...
+
+    public Timer getCheckinTimer() {
+        return _checkinTimer;
+    }
+
+    public Timer getTotalExecTimer() {
+        return _totalExecTimer;
+    }
+
+    public List<Thread> generateWorkers() {
```
Comment (milosimpson) [inline perf-test-suite/src/main/java/com/bazaarvoice/ostrich/pool/ServiceRunner.java]: /** Generates worker threads that will make requests of the ServiceCache.
Each worker thread will request a "Client" from the ServiceCache, and when it has a client will do some "busywork".   The "busywork" will be to run some cryptographic hashes across a random string.
The size of the random String to be hashed can be configured from the command line.
*/

File: perf-test-suite/README.md [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+                                  (Long.MAX_VALUE)
+
+      -s,--statistics             Output current running stats on STDOUT,
+                                  illegal if --output-file is not provided
+
+      -v,--report-every <arg>     Reports the running statistics every #
+                                  seconds
+
+
+## Interpreting the results
```
Comment (milosimpson) [inline perf-test-suite/README.md]: Need to add a "Warning" about the hotspot compiler kicking in and making things "faster" at random times.  Aka, mention the params you used to force the JVM to only JIT to a certain state.

File: perf-test-suite/pom.xml (hunk 6 above)
Comment (milosimpson) [inline perf-test-suite/pom.xml]: Also, this inclusion of the yammer metrics is contained to just this module.

File: perf-test-suite/README.md [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+    mvn -q clean compile assembly:single -Dostrich.core.version=1.9.0
+
+However, due to breaking changes between versions, i.e. 1.8 & 1.9 uses different version of metrics and
+ subsequently different constructor of ServiceCache, this 'may' require you to make changes in the test 
+ for successful compilation.
+
+
+## Running The Suite
+
+To determine the performance overhead of Ostrich under varying loads, this test suite takes various parameters
```
Comment (milosimpson) [inline perf-test-suite/README.md]: This sentence does not compile.

File: perf-test-suite/README.md [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+ for successful compilation.
+
+
+## Running The Suite
+
+To determine the performance overhead of Ostrich under varying loads, this test suite takes various parameters 
+ such as, to determine the load on the cache (# of threads), the load on each thread, etc. to determine the 
+ overhead of ostrich under nominal to mediocre to somewhat heavy loads. 
+
+After determining and setting those desired values the suite will run the desired number of thread with
```
Comment (milosimpson) [inline perf-test-suite/README.md]: thread -> threads

File: perf-test-suite/pom.xml (hunk 6 above)
Comment (ohhatiya) [inline perf-test-suite/pom.xml]: The newer metrics timer has a [getSnapshot()](https://dropwizard.github.io/metrics/3.0.2/apidocs/com/codahale/metrics/Timer.html#getSnapshot--) method which gives you [something](https://dropwizard.github.io/metrics/3.0.2/apidocs/com/codahale/metrics/Snapshot.html) with the mean and percentiles.

File: perf-test-suite/src/main/java/com/bazaarvoice/ostrich/perftest/core/SimpleServiceFactory.java [hunk 17]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    @Override
+    public void destroy(ServiceEndPoint serviceEndPoint, Service<String, String> service) {
+        service.destroy();
+    }
+
+    @Override
+    public boolean isHealthy(ServiceEndPoint endPoint) {
+        throw new RuntimeException("This should not get executed");
```
Comment (ohhatiya) [inline perf-test-suite/src/main/java/com/bazaarvoice/ostrich/perftest/core/SimpleServiceFactory.java]: Can this message be a little more descriptive? Would be nice to be able to differentiate from the one `isRetriableException` throws without having to check the stack trace.

File: perf-test-suite/src/main/java/com/bazaarvoice/ostrich/perftest/core/SimpleServiceFactory.java [hunk 18]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    @Override
+    public boolean isRetriableException(Exception exception) {
+        throw new RuntimeException("This should not get executed");
```
Comment (ohhatiya) [inline perf-test-suite/src/main/java/com/bazaarvoice/ostrich/perftest/core/SimpleServiceFactory.java]: Same comment here as above.

File: perf-test-suite/src/main/java/com/bazaarvoice/ostrich/perftest/core/SimpleServiceFactory.java [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+import com.yammer.metrics.core.TimerContext;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * A service factory, as needed in the ServiceCache
+ */
+public class SimpleServiceFactory implements ServiceFactory<Service<String, String>> {
+
+    private final Meter serviceCreated;
```
Comment (ohhatiya) [inline perf-test-suite/src/main/java/com/bazaarvoice/ostrich/perftest/core/SimpleServiceFactory.java]: Our code style uses names starting with `_` for member fields.

File: perf-test-suite/src/main/java/com/bazaarvoice/ostrich/perftest/utils/LoadRunner.java [hunk 20]
```diff
@@ -0,0 +1,164 @@
...
+import java.util.concurrent.TimeUnit;
+
+
+/**
+ * This class creates the service runner and runs the requested load as parsed via Arguments
+ * This also prints/writes the report log and/or the statistics as requested in the Arguments
+ */
+public class LoadRunner {
+
+    private final PrintStream out;
```
Comment (ohhatiya) [inline perf-test-suite/src/main/java/com/bazaarvoice/ostrich/perftest/utils/LoadRunner.java]: This fields should have names that start with `_`, too.

File: perf-test-suite/src/main/java/com/bazaarvoice/ostrich/pool/ServiceRunner.java [hunk 21]
```diff
@@ -0,0 +1,227 @@
...
+        }
+        finally {
+            _serviceMeter.mark();
+            totalTimeContext.stop();
+        }
+    }
+
+
+    /**
+     * static helpers
```
Comment (ohhatiya) [inline perf-test-suite/src/main/java/com/bazaarvoice/ostrich/pool/ServiceRunner.java]: This comment is unnecessary

File: perf-test-suite/src/main/java/com/bazaarvoice/ostrich/perftest/utils/LoadRunner.java [hunk 22]
```diff
@@ -0,0 +1,164 @@
...
+        Meter serviceDestroyed = serviceFactory.getServiceDestroyed();
+        Meter serviceCalled = serviceRunner.getServiceMeter();
+        Timer checkoutTimer = serviceRunner.getCheckoutTimer();
+        Timer checkinTimer = serviceRunner.getCheckinTimer();
+        Timer serviceTimer = serviceFactory.getServiceTimer();
+        Timer totalTimer = serviceRunner.getTotalExecTimer();
+
+        out.print(String.format("%d,", counter++));
+        out.print(String.format("%d,%.2f,%.2f,%.2f,%.2f,",
+                serviceCreated.count(), serviceCreated.meanRate(), serviceCreated.oneMinuteRate(), serviceCreated.fiveMinuteRate(), serviceCreated.fifteenMinuteRate()));
```
Comment (ohhatiya) [inline perf-test-suite/src/main/java/com/bazaarvoice/ostrich/perftest/utils/LoadRunner.java]: A lot of these print lines are too long. In other Ostrich files, we try to stick to 120 column lines.

File: perf-test-suite/pom.xml (hunk 6 above)
Comment (bivasdas) [inline perf-test-suite/pom.xml]: Thanks for the tip. We actually wanted N-percentile metrics. I've updated the latest commit with updated metrics library.

**Pull request merged**